### ISSUE TITLE
autoredraw count becomes out of sync on m.request SyntaxError

### ISSUE THREAD

**Issue by agschwender** [lhorie/mithril.js#676]
When `m.request` encounters a `SyntaxError`, it causes `m.deferred.onerror` to throw an error which causes `m.endComputation` to never fire.

```
	m.deferred.onerror = function(e) {
		if (type.call(e) === "[object Error]" && !e.constructor.toString().match(/ Error/)) throw e
	};

...

			catch (e) {
				m.deferred.onerror(e);
				deferred.reject(e)
			}
			if (xhrOptions.background !== true) m.endComputation()
```

However, looking at this test, it would seem that that behavior is intentional:

```
	test(function() {
		var error = m.prop("no error"), exception
		var prop = m.request({method: "GET", url: "test", deserialize: function() {throw new TypeError("error occurred")}}).then(null, error)
		try {mock.XMLHttpRequest.$instances.pop().onreadystatechange()}
		catch (e) {exception = e}
		m.endComputation()
		return prop() === undefined && error() === "no error" && exception.message == "error occurred"
	})
```

It seems strange to me that the caller would be expected to call `m.endComputation` when they never called `m.startComputation`. In other words, shouldn't that still be encapsulated by `m.request`? As a short term solution, I will simply override `m.deferred.onerror`, but would like to understand the correct behavior here as a caller.

**Comment by lawrence-dol** [comment]
This kind of thing is why the start/endComputation is error-prone and should be avoided. Instead, bits of the UI that are waiting on network requests should either not draw or show some sort of "pending" indication, and redrawing should never be blocked.

**Comment by barneycarroll** [comment]
@lawrence-dol totally agree, it's way over exposed for an API of dubious value that (as @agschwender has shown) requires you to write an asynchronous try finally loop in order to run safely.

Totally support this fix (which fixes internal use), but would recommend depreciation in the docs. Blocking XHR allows you to write extremely simple views with incredibly little code, but even then blocking UIs are an anti-pattern in modern web design. Personally I can't think of a single good use case for which users might want this footgun, and yet a lot of code I see in support requests has people using this API when they could be using plain redraw calls.

**Comment by lhorie** [comment]
@agschwender This is intended behavior.

In a nutshell, the promises spec has a design flaw which ultimately results in errors being swallowed by default, instead of throwing and being reported in console as other errors in js. Mithril uses an exception monitor (an idea similar to Ember's) to provide more sensible error messages for native errors in async contexts, while letting app devs throw app errors via the Error class. 

You can override the exception monitor as documented here: http://lhorie.github.io/mithril/mithril.deferred.html#the-exception-monitor, but the default behavior is indeed that throwing an error during a redraw breaks redrawing.

**Comment by agschwender** [comment]
I think the issue for me is not that that the error itself stops that particular redraw, but that it stops all future auto-redraws by not having decremented the pendingRequests count. Basically, it leaves the application in a completely frozen, non-responsive state. In an application/database sense which is more of my background, it would be like leaving the database state in a broken/partial state rather than executing a rollback on error. Leaving it to the caller to correct the internal state after an error (as that test would imply) seems to be a little fragile in and of itself, as the caller would need to be constantly aware of the internals of m.request in order to safely catch the error, restore the state and proceed as normal. 

In any case, will continue to override m.deferred.onerror to perform a noop.

**Comment by lawrence-dol** [comment]
@lhorie, The problem is not *due* to a fault in the promises spec, it's *exposed* by it. The reality is that start/endComputation is error-prone and brittle, and it's unnecessary to boot because it's simply bad practice to block rendering while network requests are pending. It there is one API that I think needs to be withdrawn in Mithril, this is it.

**Comment by tropperstyle** [comment]
Surprised to see so much knocking of the start/endCompuration API. As someone who has had to deal with a whole lot of async code, the computation API has been a godsend in optimizing redraws. Though this definitely seems like a bug with pendingRequest not being decremented....

**Comment by lhorie** [comment]
FWIW, I added a line of code to reset the counter to zero when the exception monitor throws, so this should no longer be an issue

**Comment by tropperstyle** [comment]
I wrap all my API calls in a module that uses the maybeJSON pattern. Essentially all app requests use:
```javascript
m.request({ ..., deserialize: maybeJSON });

function maybeJSON(data) {
  try { return JSON.parse(data) }
  catch (e) { throw new Error('Invalid Server Response') }
}
```

This must be how I worked around this exact issue. I was trying to find the related blog post or documentation where I found the code but had some trouble until I found @lhorie's comment in [#86](https://github.com/lhorie/mithril.js/issues/86#issuecomment-43707634)

I understand the logic to break redrawing on non-recoverable errors, as they will always require code changes to be fixed. So the question then is should the default serializer encapsulate this behavior and de-escalate the error to application space? At the very least I think the maybeJSON pattern needs to be in official documentation.
